File: advanced-api-design/02-intermediate/src/gateway/typescript/gateway.ts
/**
 * Minimal API Gateway:
 * - Reverse proxy routing
 * - Token-bucket rate limiting per API key
 * - Request transformation (strip Authorization → X-User-Id)
 * - Simple JWT-like demo token (base64 payload) — NOT for production crypto
 *
 * Upstream mock: run `npx tsx upstream.ts` on :5001
 * Gateway: npx tsx gateway.ts on :8080
 */
import express, { Request, Response, NextFunction } from 'express';
import { createProxyMiddleware } from 'http-proxy-middleware';
import { randomUUID } from 'node:crypto';

const PORT = Number(process.env.PORT) || 8080;
const UPSTREAM = process.env.UPSTREAM || 'http://127.0.0.1:5001';

interface Bucket {
  tokens: number;
  updatedAt: number;
}

const buckets = new Map<string, Bucket>();
const CAPACITY = 10;
const REFILL_PER_SEC = 2;

function takeToken(key: string): boolean {
  const now = Date.now();
  let b = buckets.get(key);
  if (!b) {
    b = { tokens: CAPACITY, updatedAt: now };
    buckets.set(key, b);
  }
  const elapsed = (now - b.updatedAt) / 1000;
  b.tokens = Math.min(CAPACITY, b.tokens + elapsed * REFILL_PER_SEC);
  b.updatedAt = now;
  if (b.tokens < 1) return false;
  b.tokens -= 1;
  return true;
}

/** Demo token: "user:<id>" base64 — replace with real JWT verify in production */
function parseDemoToken(auth?: string): { userId: string } | null {
  if (!auth?.startsWith('Bearer ')) return null;
  try {
    const raw = Buffer.from(auth.slice(7), 'base64').toString('utf8');
    const [kind, userId] = raw.split(':');
    if (kind !== 'user' || !userId) return null;
    return { userId };
  } catch {
    return null;
  }
}

function authAndLimit(req: Request, res: Response, next: NextFunction) {
  const identity = parseDemoToken(req.header('authorization') ?? undefined);
  if (!identity) {
    return res.status(401).type('application/problem+json').json({
      title: 'Unauthorized',
      status: 401,
      detail: "Bearer token required (demo: base64 of 'user:<id>')",
    });
  }

  const key = identity.userId;
  if (!takeToken(key)) {
    res.setHeader('Retry-After', '1');
    return res.status(429).type('application/problem+json').json({
      title: 'Too Many Requests',
      status: 429,
      detail: 'rate limit exceeded',
    });
  }

  // Transformation: never forward raw Authorization to internal services
  req.headers['x-user-id'] = identity.userId;
  req.headers['x-request-id'] = req.header('x-request-id') || randomUUID();
  delete req.headers['authorization'];
  next();
}

const app = express();

app.get('/health', (_req, res) => res.json({ status: 'ok', role: 'gateway' }));

// External path /v1/orders/* → internal /orders/*
app.use(
  '/v1/orders',
  authAndLimit,
  createProxyMiddleware({
    target: UPSTREAM,
    changeOrigin: true,
    pathRewrite: { '^/v1/orders': '/orders' },
    on: {
      proxyReq: (proxyReq, req) => {
        const r = req as Request;
        if (r.headers['x-user-id']) proxyReq.setHeader('X-User-Id', String(r.headers['x-user-id']));
        if (r.headers['x-request-id'])
          proxyReq.setHeader('X-Request-Id', String(r.headers['x-request-id']));
      },
    },
  }),
);

app.listen(PORT, () => {
  const demo = Buffer.from('user:alice').toString('base64');
  console.log(`API Gateway on :${PORT} → ${UPSTREAM}`);
  console.log(`Demo token: Bearer ${demo}`);
  console.log(`Try: curl -H "Authorization: Bearer ${demo}" http://localhost:${PORT}/v1/orders`);
});

File: advanced-api-design/02-intermediate/src/gateway/python/gateway.py
"""
API Gateway (Python / Flask): reverse proxy + token bucket + header transform
Requires upstream on :5001 (use the TypeScript upstream.ts or any mock).
"""
from __future__ import annotations

import base64
import time
import uuid
from threading import Lock

import requests
from flask import Flask, Response, jsonify, request

app = Flask(__name__)
UPSTREAM = "http://127.0.0.1:5001"
CAPACITY = 10.0
REFILL = 2.0

_buckets: dict[str, dict] = {}
_lock = Lock()


def take_token(key: str) -> bool:
    now = time.time()
    with _lock:
        b = _buckets.get(key)
        if not b:
            b = {"tokens": CAPACITY, "updated": now}
            _buckets[key] = b
        elapsed = now - b["updated"]
        b["tokens"] = min(CAPACITY, b["tokens"] + elapsed * REFILL)
        b["updated"] = now
        if b["tokens"] < 1:
            return False
        b["tokens"] -= 1
        return True


def parse_demo_token(auth: str | None) -> str | None:
    if not auth or not auth.startswith("Bearer "):
        return None
    try:
        raw = base64.b64decode(auth[7:]).decode()
        kind, user_id = raw.split(":", 1)
        if kind != "user" or not user_id:
            return None
        return user_id
    except Exception:
        return None


@app.get("/health")
def health():
    return jsonify(status="ok", role="gateway")


@app.route("/v1/orders", defaults={"path": ""}, methods=["GET", "POST", "PATCH", "PUT", "DELETE", "OPTIONS"])
@app.route("/v1/orders/<path:path>", methods=["GET", "POST", "PATCH", "PUT", "DELETE", "OPTIONS"])
def proxy_orders(path: str):
    if request.method == "OPTIONS":
        resp = Response()
        resp.headers["Access-Control-Allow-Origin"] = "*"
        resp.headers["Access-Control-Allow-Methods"] = "GET, POST, PATCH, PUT, DELETE, OPTIONS"
        resp.headers["Access-Control-Allow-Headers"] = "Authorization, Content-Type, X-Request-Id"
        resp.headers["Access-Control-Max-Age"] = "86400"
        return resp
    user_id = parse_demo_token(request.headers.get("Authorization"))
    if not user_id:
        return jsonify(title="Unauthorized", status=401, detail="Bearer token required"), 401
    if not take_token(user_id):
        resp = jsonify(title="Too Many Requests", status=429, detail="rate limit exceeded")
        resp.status_code = 429
        resp.headers["Retry-After"] = "1"
        return resp

    target = f"{UPSTREAM}/orders" + (f"/{path}" if path else "")
    if request.query_string:
        target += "?" + request.query_string.decode()

    headers = {
        "X-User-Id": user_id,
        "X-Request-Id": request.headers.get("X-Request-Id") or str(uuid.uuid4()),
        "Content-Type": request.headers.get("Content-Type", "application/json"),
    }
    upstream = requests.request(
        method=request.method,
        url=target,
        headers=headers,
        data=request.get_data(),
        timeout=5,
    )
    return Response(upstream.content, status=upstream.status_code, content_type=upstream.headers.get("Content-Type"))


if __name__ == "__main__":
    demo = base64.b64encode(b"user:alice").decode()
    print("API Gateway (Python) on :8080")
    print(f"Demo token: Bearer {demo}")
    app.run(port=8080)
